controllers/nugus_controller/nugus_controller.cpp

```diff
@@ -15,107 +15,452 @@
-            send(tcp_fd, "Welcome", 8, 0);
+        send(tcp_fd, "Welcome", 8, 0);
     }
     ~NUgus() {
         close_socket(tcp_fd);
     }
 
     void run() {
-        // Message counter
-        uint32_t current_num = 1;
+        int controller_time = 0;
 
         while (step(time_step) != -1) {
             // Don't bother doing anything unless we have an active TCP connection
             if (tcp_fd == -1) {
                 std::cerr << "Error: Failed to start TCP server, retrying ..." << std::endl;
                 tcp_fd = create_socket_server(server_port);
                 send(tcp_fd, "Welcome", 8, 0);
+                controller_time = 0;
                 continue;
             }
+            controller_time += this->getBasicTimeStep();
 
-            // Setup arguments for select call
-            fd_set rfds;
-            FD_ZERO(&rfds);
-            FD_SET(tcp_fd, &rfds);
-            timeval timeout = {0, 0};
+            // Send data from the simulated robot hardware to the robot control software
+            sendData(controller_time);
 
-            // Watch TCP file descriptor to see when it has input.
-            // No wait - polling as fast as possible
-            int num_ready = select(tcp_fd + 1, &rfds, nullptr, nullptr, &timeout);
-            if (num_ready < 0) {
-                std::cerr << "Error: Polling of TCP connection failed: " << strerror(errno) << std::endl;
+            // Check if we have received a message and deal with it if we have
+            try {
+                handleReceived();
+            }
+            catch (const std::exception& e) {
+                std::cout << "didn't work" << std::endl;
                 continue;
             }
-            else if (num_ready > 0) {
-                // Wire format
-                // unit32_t Nn  message size in bytes. The bytes are in network byte order (big endian)
-                // uint8_t * Nn  the message
-                uint32_t Nn;
-                if (recv(tcp_fd, &Nn, sizeof(Nn), 0) != sizeof(Nn)) {
-                    std::cerr << "Error: Failed to read message size from TCP connection: " << strerror(errno)
-                              << std::endl;
-                    continue;
+        }
+    }
+
+    void handleReceived() {
+        // Setup arguments for select call
+        fd_set rfds;
+        FD_ZERO(&rfds);
+        FD_SET(tcp_fd, &rfds);
+        timeval timeout = {0, 0};
+
+        // Watch TCP file descriptor to see when it has input.
+        // No wait - polling as fast as possible
+        int num_ready = select(tcp_fd + 1, &rfds, nullptr, nullptr, &timeout);
+        if (num_ready < 0) {
+            std::cerr << "Error: Polling of TCP connection failed: " << strerror(errno) << std::endl;
+            throw;
+        }
+        else if (num_ready > 0) {
+            // Wire format
+            // unit32_t Nn  message size in bytes. The bytes are in network byte order (big endian)
```

Comment:
If `Nn` is zero, the likely corresponds to a protobuf message with all values set to their defaults

Reply:
When you say defaults, do you mean empty? Is it the same thing as just leaving it as `ActuatorRequests actuatorRequests;`?

Reply:
It should be the same as that